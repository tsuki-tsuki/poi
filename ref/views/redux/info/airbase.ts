import { trimArray, compareUpdate, constructArray } from 'views/utils/tools'
import { zip, findIndex, get, map, omit, unzip } from 'lodash'
import { createSlice } from '@reduxjs/toolkit'
import { APIAirBase } from 'kcsapi/api_get_member/mapinfo/response'

import {
  createAPIGetMemberMapinfoResponseAction,
  createAPIReqAirCorpsSetPlaneResponseAction,
  createAPIReqAirCorpsChangeNameResponseAction,
  createAPIReqAirCorpsSetActionResponseAction,
  createAPIReqAirCorpsSupplyResponseAction,
  createAPIReqMapNextResponseAction,
  createAPIPortPortResponseAction,
  createAPIReqAirCorpsChangeDeploymentBaseResponseAction,
  APIBaseItem,
} from '../actions'

export interface AirBase extends Partial<APIAirBase> {
  api_maxhp?: number
  api_nowhp?: number
}

const airBaseSlice = createSlice({
  name: 'airbase',
  initialState: [] as AirBase[],
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(createAPIGetMemberMapinfoResponseAction, (state, { payload }) => {
        const airbase = payload.body.api_air_base
        if (typeof airbase === 'undefined') {
          return state
        }
        return trimArray(compareUpdate(state, airbase, 3), airbase)
      })
      .addCase(createAPIReqAirCorpsSetPlaneResponseAction, (state, { payload }) => {
        const { api_base_id, api_area_id } = payload.postBody
        const { api_distance, api_plane_info } = payload.body
        const baseIndex = findIndex(
          state,
          (squad) => squad.api_rid === +api_base_id && squad.api_area_id === +api_area_id,
        )
        const index = baseIndex === -1 ? +api_base_id - 1 : baseIndex
        const squadrons = constructArray(
          api_plane_info.map((p) => p.api_squadron_id - 1),
          api_plane_info,
        )
        return compareUpdate(
          state,
          constructArray(
            [index],
            [
              {
                api_distance,
                api_plane_info: squadrons,
              },
            ],
          ),
          3,
        )
      })
      .addCase(createAPIReqAirCorpsChangeDeploymentBaseResponseAction, (state, { payload }) => {
        const { api_area_id, api_base_id_src, api_base_id, api_item_id } = payload.postBody
        const { api_base_items } = payload.body
        // Err on the side of caution, few preconditions before updating.

        // Only two plane slots are supposed to be swapped.
        if (api_base_items.length !== 2) {
          return state
        }

        const findSquadronIndex = (baseId: number | string) => {
          const ret = findIndex(
            state,
            (squad) => squad.api_rid === +baseId && squad.api_area_id === +api_area_id,
          )
          return ret === -1 ? +baseId - 1 : ret
        }

        // Moving a plane equip from src slot to dst slot.
        const indexSrc = findSquadronIndex(api_base_id_src)
        const indexDst = findSquadronIndex(api_base_id)

        // The equip in question should exist.
        if (
          findIndex(
            state[indexSrc].api_plane_info,
            (squad) => squad.api_slotid === +api_item_id,
          ) === -1
        ) {
          return state
        }

        const [objSrc, objDst] = (() => {
          const [item0, item1] = api_base_items
          return item0.api_rid === +api_base_id_src ? [item0, item1] : [item1, item0]
        })()

        const convertItem = ({ api_distance, api_plane_info }: APIBaseItem) => {
          const squadrons = constructArray(
            api_plane_info.map((p) => p.api_squadron_id - 1),
            api_plane_info,
          )
          return {
            api_distance,
            api_plane_info: squadrons,
          }
        }

        return compareUpdate(
          state,
          constructArray([indexSrc, indexDst], map([objSrc, objDst], convertItem)),
          3,
        )
      })
      .addCase(createAPIReqAirCorpsChangeNameResponseAction, (state, { payload }) => {
        const { api_base_id, api_name, api_area_id } = payload.postBody
        const baseIndex = findIndex(
          state,
          (squad) => squad.api_rid === +api_base_id && squad.api_area_id === +api_area_id,
        )
        const index = baseIndex === -1 ? +api_base_id - 1 : baseIndex
        return compareUpdate(
          state,
          constructArray(
            [index],
            [
              {
                api_name,
              },
            ],
          ),
          2,
        )
      })
      .addCase(createAPIReqAirCorpsSetActionResponseAction, (state, { payload }) => {
        const { api_action_kind, api_base_id, api_area_id } = payload.postBody
        const update = (
          zip(api_base_id.split(','), api_action_kind.split(',')) as [string, string][]
        ).map(([base_id, action_kind]) => {
          const baseIndex = findIndex(
            state,
            (squad) => squad.api_rid === +base_id && squad.api_area_id === +api_area_id,
          )
          const index = baseIndex === -1 ? +base_id - 1 : baseIndex
          return [index, { api_action_kind: parseInt(action_kind) }]
        })
        const [idx, values] = unzip(update) as [number[], Partial<APIAirBase>[]]
        return compareUpdate(state, constructArray(idx, values), 2)
      })
      .addCase(createAPIReqAirCorpsSupplyResponseAction, (state, { payload }) => {
        const { api_base_id, api_area_id } = payload.postBody
        const { api_plane_info } = payload.body
        const baseIndex = findIndex(
          state,
          (squad) => squad.api_rid === +api_base_id && squad.api_area_id === +api_area_id,
        )
        const index = baseIndex === -1 ? +api_base_id - 1 : baseIndex
        const squadrons = constructArray(
          api_plane_info.map((p) => p.api_squadron_id - 1),
          api_plane_info,
        )
        return compareUpdate(
          state,
          constructArray(
            [index],
            [
              {
                api_plane_info: squadrons,
              },
            ],
          ),
          3,
        )
      })
      .addCase(createAPIReqMapNextResponseAction, (state, { payload }) => {
        const { api_destruction_battle, api_maparea_id } = payload.body
        if (api_destruction_battle) {
          const { api_f_maxhps, api_f_nowhps, api_air_base_attack } = api_destruction_battle
          const parsed_api_air_base_attack =
            typeof api_air_base_attack === 'string'
              ? JSON.parse(api_air_base_attack)
              : api_air_base_attack
          const api_fdam = get(parsed_api_air_base_attack, 'api_stage3.api_fdam', [])
          return map(state, (airbase) => {
            const { api_area_id, api_rid } = airbase
            if (api_maparea_id !== api_area_id) {
              return airbase
            }

            const index = (api_rid || 0) - 1
            const newBase = { ...airbase }

            if (get(api_f_maxhps, index) >= 0) {
              newBase.api_maxhp = api_f_maxhps[index]
            }

            if (get(api_f_nowhps, index) >= 0) {
              newBase.api_nowhp = api_f_nowhps[index] - get(api_fdam, index, 0)
            }

            return newBase
          })
        }
      })
      .addCase(createAPIPortPortResponseAction, (state) => {
        return map(state, (airbase) =>
          typeof airbase.api_nowhp !== 'undefined' || typeof airbase.api_maxhp !== 'undefined'
            ? omit(airbase, ['api_nowhp', 'api_maxhp'])
            : airbase,
        )
      })
  },
})

export const reducer = airBaseSlice.reducer
