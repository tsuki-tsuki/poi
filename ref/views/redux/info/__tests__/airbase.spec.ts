import { reducer } from '../airbase'

// FIXME: cannot find a good fixture for createAPIReqMapNextResponseAction

import {
  createAPIGetMemberMapinfoResponseAction,
  createAPIReqAirCorpsSetPlaneResponseAction,
  createAPIReqAirCorpsChangeNameResponseAction,
  createAPIReqAirCorpsSetActionResponseAction,
  createAPIReqAirCorpsSupplyResponseAction,
  createAPIPortPortResponseAction,
} from '../../actions'

import mapInfoFixture from './__fixtures__/api_get_member_mapinfo.json'
import setPlaneFixture from './__fixtures__/api_req_air_corps_set_plane.json'
import changeNameFixture from './__fixtures__/api_req_air_corps_change_name.json'
import setActionFixture from './__fixtures__/api_req_air_corps_set_action.json'
import supplyFixture from './__fixtures__/api_req_air_corps_supply.json'

describe('airbase reduer', () => {
  const initialState = reducer([], createAPIGetMemberMapinfoResponseAction(mapInfoFixture))

  it('empty action', () => {
    // @ts-expect-error testing empty reducer
    expect(reducer(undefined, {})).toEqual([])
  })

  it('createAPIGetMemberMapinfoResponseAction', () => {
    // @ts-expect-error testing empty reducer
    expect(reducer([], createAPIGetMemberMapinfoResponseAction({ body: {} }))).toEqual([])

    expect(reducer([], createAPIGetMemberMapinfoResponseAction(mapInfoFixture))).toMatchSnapshot()
  })

  it('createAPIReqAirCorpsSetPlaneResponseAction', () => {
    expect(initialState).toMatchDiffSnapshot(
      reducer(initialState, createAPIReqAirCorpsSetPlaneResponseAction(setPlaneFixture)),
    )
  })

  it('createAPIReqAirCorpsChangeNameResponseAction', () => {
    expect(initialState).toMatchDiffSnapshot(
      reducer(initialState, createAPIReqAirCorpsChangeNameResponseAction(changeNameFixture)),
    )
  })

  it('createAPIReqAirCorpsSetActionResponseAction', () => {
    expect(initialState).toMatchDiffSnapshot(
      reducer(initialState, createAPIReqAirCorpsSetActionResponseAction(setActionFixture)),
    )
  })

  it('createAPIReqAirCorpsSupplyResponseAction', () => {
    expect(initialState).toMatchDiffSnapshot(
      reducer(initialState, createAPIReqAirCorpsSupplyResponseAction(supplyFixture)),
    )
  })

  it('createAPIPortPortResponseAction', () => {
    expect(initialState).toMatchDiffSnapshot(
      // @ts-expect-error testing empty reducer
      reducer(initialState, createAPIPortPortResponseAction({})),
    )
  })
})
