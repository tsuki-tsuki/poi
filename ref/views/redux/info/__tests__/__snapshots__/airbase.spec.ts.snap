// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`airbase reduer createAPIGetMemberMapinfoResponseAction 1`] = `
Array [
  Object {
    "api_action_kind": 1,
    "api_area_id": 6,
    "api_distance": Object {
      "api_base": 5,
      "api_bonus": 0,
    },
    "api_name": "第一基地航空隊",
    "api_plane_info": Array [
      Object {
        "api_cond": 1,
        "api_count": 18,
        "api_max_count": 18,
        "api_slotid": 26059,
        "api_squadron_id": 1,
        "api_state": 1,
      },
      Object {
        "api_cond": 1,
        "api_count": 4,
        "api_max_count": 18,
        "api_slotid": 52364,
        "api_squadron_id": 2,
        "api_state": 1,
      },
      Object {
        "api_cond": 1,
        "api_count": 15,
        "api_max_count": 18,
        "api_slotid": 52792,
        "api_squadron_id": 3,
        "api_state": 1,
      },
      Object {
        "api_cond": 1,
        "api_count": 17,
        "api_max_count": 18,
        "api_slotid": 52691,
        "api_squadron_id": 4,
        "api_state": 1,
      },
    ],
    "api_rid": 1,
  },
  Object {
    "api_action_kind": 0,
    "api_area_id": 6,
    "api_distance": Object {
      "api_base": 0,
      "api_bonus": 0,
    },
    "api_name": "第二基地航空隊",
    "api_plane_info": Array [
      Object {
        "api_slotid": 0,
        "api_squadron_id": 1,
        "api_state": 0,
      },
      Object {
        "api_slotid": 0,
        "api_squadron_id": 2,
        "api_state": 0,
      },
      Object {
        "api_slotid": 0,
        "api_squadron_id": 3,
        "api_state": 0,
      },
      Object {
        "api_slotid": 0,
        "api_squadron_id": 4,
        "api_state": 0,
      },
    ],
    "api_rid": 2,
  },
  Object {
    "api_action_kind": 0,
    "api_area_id": 6,
    "api_distance": Object {
      "api_base": 0,
      "api_bonus": 0,
    },
    "api_name": "第三基地航空隊",
    "api_plane_info": Array [
      Object {
        "api_slotid": 0,
        "api_squadron_id": 1,
        "api_state": 0,
      },
      Object {
        "api_slotid": 0,
        "api_squadron_id": 2,
        "api_state": 0,
      },
      Object {
        "api_slotid": 0,
        "api_squadron_id": 3,
        "api_state": 0,
      },
      Object {
        "api_slotid": 0,
        "api_squadron_id": 4,
        "api_state": 0,
      },
    ],
    "api_rid": 3,
  },
]
`;

exports[`airbase reduer createAPIPortPortResponseAction 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`airbase reduer createAPIReqAirCorpsChangeNameResponseAction 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -4,11 +4,11 @@
      \\"api_area_id\\": 6,
      \\"api_distance\\": Object {
        \\"api_base\\": 5,
        \\"api_bonus\\": 0,
      },
-     \\"api_name\\": \\"第一基地航空隊\\",
+     \\"api_name\\": \\"第1基地航空隊\\",
      \\"api_plane_info\\": Array [
        Object {
          \\"api_cond\\": 1,
          \\"api_count\\": 18,
          \\"api_max_count\\": 18,"
`;

exports[`airbase reduer createAPIReqAirCorpsSetActionResponseAction 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,8 +1,8 @@
  Array [
    Object {
-     \\"api_action_kind\\": 1,
+     \\"api_action_kind\\": 0,
      \\"api_area_id\\": 6,
      \\"api_distance\\": Object {
        \\"api_base\\": 5,
        \\"api_bonus\\": 0,
      },"
`;

exports[`airbase reduer createAPIReqAirCorpsSetPlaneResponseAction 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -8,13 +8,13 @@
      },
      \\"api_name\\": \\"第一基地航空隊\\",
      \\"api_plane_info\\": Array [
        Object {
          \\"api_cond\\": 1,
-         \\"api_count\\": 18,
+         \\"api_count\\": 15,
          \\"api_max_count\\": 18,
-         \\"api_slotid\\": 26059,
+         \\"api_slotid\\": 52792,
          \\"api_squadron_id\\": 1,
          \\"api_state\\": 1,
        },
        Object {
          \\"api_cond\\": 1,
@@ -24,13 +24,13 @@
          \\"api_squadron_id\\": 2,
          \\"api_state\\": 1,
        },
        Object {
          \\"api_cond\\": 1,
-         \\"api_count\\": 15,
+         \\"api_count\\": 18,
          \\"api_max_count\\": 18,
-         \\"api_slotid\\": 52792,
+         \\"api_slotid\\": 26059,
          \\"api_squadron_id\\": 3,
          \\"api_state\\": 1,
        },
        Object {
          \\"api_cond\\": 1,"
`;

exports[`airbase reduer createAPIReqAirCorpsSupplyResponseAction 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -24,11 +24,11 @@
          \\"api_squadron_id\\": 2,
          \\"api_state\\": 1,
        },
        Object {
          \\"api_cond\\": 1,
-         \\"api_count\\": 15,
+         \\"api_count\\": 18,
          \\"api_max_count\\": 18,
          \\"api_slotid\\": 52792,
          \\"api_squadron_id\\": 3,
          \\"api_state\\": 1,
        },"
`;
